---
title: "Creating features for every genome"
author: "Yva Sandvik"
date: "1/14/2022"
output: html_document
---

# Installer pakker

```{r}
library(remotes)
#install_github('larssnip/microseq', 'translate-codon', force = TRUE)
#install_github('larssnip/microclass', 'in-frame-K-mer', force = TRUE)
library(microclass)
library(microseq)
library(seqinr)
```

# Load ORF-dataset for a genome

```{r}

# Denne koden må justeres ut i fra hvilket genom man prøver å prosessere

load("~/Documents/NMBU/Semester 12/Data Science Master/data-science-thesis/data_expo/G1_orfs.RData")


G.dataset <- subset(G1.dataset, select=c(Length, Sequence))

```

# Compute GC-content features

```{r}
GC_features <- function(data){
  data %>%
    select(Length, Sequence) %>%
    mutate(Sequence = strsplit(Sequence, '')) -> new.tbl
  GC <- c(1:nrow(new.tbl))
  GC1 <- c(1:nrow(new.tbl))
  GC2 <- c(1:nrow(new.tbl))
  GC3 <- c(1:nrow(new.tbl))
  for(i in GC){
    seq <- unlist(new.tbl$Sequence[i])
    GC[i] <- GC(seq)
    GC1[i] <- GC1(seq)
    GC2[i] <- GC2(seq)
    GC3[i] <- GC3(seq)
  }
  result <- data.frame(GC_content = GC, GC1_content = GC1, GC2_content = GC2, GC3_content = GC3)
  return(result)
}
```

```{r}
GC_features <- GC_features(G.dataset)
```

# Computing 2-mer features 

```{r}
# Returns a matrix of K-mer counts given a dataset
Kmer_matrix <- KmerCount(G.dataset$Sequence, K=2, col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_kmer <- Kmer_matrix/rowSums(Kmer_matrix)

# Coverting from matrix to to data frame
Kmer_2 <- as.data.frame(standardised_kmer)
```

# Computing 3-mer features 

```{r}
# Returns a matrix of K-mer counts given a dataset
Kmer_matrix <- KmerCount(G.dataset$Sequence, K=3, col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_kmer <- Kmer_matrix/rowSums(Kmer_matrix)

# Coverting from matrix to to data frame
Kmer_3 <- as.data.frame(standardised_kmer)
```

# Computing 4-mer features 

```{r}
# Returns a matrix of K-mer counts given a dataset
Kmer_matrix <- KmerCount(G.dataset$Sequence, K=4, col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_kmer <- Kmer_matrix/rowSums(Kmer_matrix)

# Coverting from matrix to to data frame
Kmer_4 <- as.data.frame(standardised_kmer)
```

# Computing 5-mer features 

```{r}
# Returns a matrix of K-mer counts given a dataset
Kmer_matrix <- KmerCount(G.dataset$Sequence, K=5, col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_kmer <- Kmer_matrix/rowSums(Kmer_matrix)

# Coverting from matrix to to data frame
Kmer_5 <- as.data.frame(standardised_kmer)
```

# Computing 6-mer features 

```{r}
# Returns a matrix of K-mer counts given a dataset
Kmer_matrix <- KmerCount(G.dataset$Sequence, K=6, col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_kmer <- Kmer_matrix/rowSums(Kmer_matrix)

# Coverting from matrix to to data frame
Kmer_6 <- as.data.frame(standardised_kmer)
```

# Computing aminoacid-1-mer features

```{r}

# Returns a matrix of K-mer counts given a dataset
aa_mer_matrix <- KmerCount(G.dataset$Sequence, K=1, type = 'amino acid', col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_aa_mer <- aa_mer_matrix/rowSums(aa_mer_matrix)

# Coverting from matrix to to data frame
aa_mer_1 <- as.data.frame(standardised_aa_mer)

```

# Computing aminoacid-2-mer features

```{r}

# Returns a matrix of K-mer counts given a dataset
aa_mer_matrix <- KmerCount(G.dataset$Sequence, K=2, type = 'amino acid', col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_aa_mer <- aa_mer_matrix/rowSums(aa_mer_matrix)

# Coverting from matrix to to data frame
aa_mer_2 <- as.data.frame(standardised_aa_mer)

```

# Computing aminoacid-3-mer features

```{r}

# Returns a matrix of K-mer counts given a dataset
aa_mer_matrix <- KmerCount(G.dataset$Sequence, K=3, type = 'amino acid', col.names = TRUE)

# Dividing all K-mer counts in each sequence by sequence length 
standardised_aa_mer <- aa_mer_matrix/rowSums(aa_mer_matrix)

# Coverting from matrix to to data frame
aa_mer_3 <- as.data.frame(standardised_aa_mer)

```

# Computing codon bias feature (c_weight)

```{r}

# Function 1

c_weight <- function(seq){
  # get codon count and amino acids count for the sequence
  seq <- substring(seq, 4)
  cc <- KmerCount(seq, K = 3, col.names = TRUE, type = 'nucleotide 1')
  aa <- tapply(cc, microseq::translate(colnames(cc), no.stop=FALSE, M.start = FALSE), sum)
  dim(aa) <- c(1,21)
  colnames(aa) <- c("*", "A", "C", "D", "E", "F", "G", "H", "I", "K", "L", "M", "N", "P", "Q",
                    "R", "S", "T", "V", "W", "Y")
  
  # translate the column names for the codon count vector and save as a copied dataframe
  trans_col <- microseq::translate(colnames(cc), M.start = FALSE)
  cc_copy <- cc
  colnames(cc_copy) <- trans_col
  
  # match the two dataframes and save the indexes
  index <- match(colnames(cc_copy), colnames(aa))
  index[is.na(index)] <- 1

  # loop through the vector and divide the codon count by the respective amino acid count
  for(i in c(1:64)){
      codon_count <- cc_copy[i]
      aa_count <- aa[index[i]]
      c_weight <- codon_count/aa_count
      cc_copy[i] <- c_weight
  }
  # Remove na values and return codon count 
  # c_copy[49]
  cc_copy[is.na(cc_copy)] <- 0
  colnames(cc_copy) <- colnames(cc)
  return(cc_copy)
}

# Function 2

c_weight.tbl <- function(data){
  cw_features <- c_weight(data$Sequence[1])
  for(i in c(2:nrow(data))){
    weights <- c_weight(data$Sequence[i])
    cw_features <- rbind(cw_features, weights)
  }
  return(cw_features)
}
```

```{r}

c_weights <- c_weight.tbl(G.dataset)
c_weights <- as.data.frame(c_weights)

colnames(c_weights) <- c("AAA_c_weight", "AAC_c_weight", "AAG_c_weight", "AAT_c_weight", "ACA_c_weight", "ACC_c_weight", "ACG_c_weight", "ACT_c_weight", "AGA_c_weight", "AGC_c_weight", "AGG_c_weight", "AGT_c_weight", "ATA_c_weight", "ATC_c_weight", "ATG_c_weight", "ATT_c_weight", "CAA_c_weight", "CAC_c_weight", "CAG_c_weight", "CAT_c_weight", "CCA_c_weight", "CCC_c_weight", "CCG_c_weight", "CCT_c_weight", "CGA_c_weight", "CGC_c_weight", "CGG_c_weight", "CGT_c_weight", "CTA_c_weight", "CTC_c_weight", "CTG_c_weight", "CTT_c_weight", "GAA_c_weight", "GAC_c_weight", "GAG_c_weight", "GAT_c_weight", "GCA_c_weight", "GCC_c_weight", "GCG_c_weight", "GCT_c_weight", "GGA_c_weight", "GGC_c_weight", "GGG_c_weight", "GGT_c_weight", "GTA_c_weight", "GTC_c_weight", "GTG_c_weight", "GTT_c_weight", "TAA_c_weight", "TAC_c_weight", "TAG_c_weight", "TAT_c_weight", "TCA_c_weight", "TCC_c_weight", "TCG_c_weight", "TCT_c_weight", "TGA_c_weight", "TGC_c_weight", "TGG_c_weight", "TGT_c_weight", "TTA_c_weight", "TTC_c_weight", "TTG_c_weight", "TTT_c_weight")

```

# Computing start codon feature

```{r}

start_codon <- function(data){
  start <- matrix(nrow=nrow(data), ncol=3)
  colnames(start) <- c("Start_ATG", "Start_GTG", "Start_TTG")
  for(i in c(1:nrow(data))){
    seq <- substring(data$Sequence[i], 1,3)
    if(seq == 'ATG'){
      s <- c(1,0,0)
      start[i, 1:3] <- s
    } else if(seq == 'GTG'){
      s <- c(0,1,0)
      start[i, 1:3] <- s
    } else if(seq == 'TTG'){
      s <- c(0,0,1)
      start[i, 1:3] <- s
    } else {
      s <- c(0,0,0)
      start[i, 1:3] <- s
    }
  }
  return(start <-  as.data.frame(start))
}

```

```{r}
Start_features <- start_codon(G.dataset)
```

# Combining all column features to form new feature dataset

```{r}

# Denne koden må justeres ut i fra hvilket genom man prøver å prosessere

G_features <- bind_cols(G15.dataset, GC_features, Start_features, Kmer_2, Kmer_3, Kmer_4, Kmer_5, Kmer_6, c_weights)

# Need genome column, type column, and feature columns, the rest is not necessary 
G_features <- subset(G_features, select=-c(Sequence, Category, Signature))
```

# Save features for given genome

```{r}

# Denne koden må justeres ut i fra hvilket genom man prøver å prosessere

G1.features <- G_features

# RData
save(G1.features, file="G1_orf_features.RData")

# CSV
write.csv(G1.features, file="G1.features.csv")

```

